server/lib/pbench/process_tb.py

```diff
@@ -76,61 +94,305 @@ def _results_push(controller: str, tb: Path, token: str):
+            # 2. Verifies the tar ball's MD5 signature
+            # 3. Creates all the link destination directory symlinks
+
+            # Combined operations will do
+            # 1. PUT API if have a token
+            # 2. p1
+            # 3. d1
+            # 4. d2
+            # 5. p3
+            # 6. p4
+            # 7. p5
+            # 8. p6
+            # 9. d3
+
+            # Extracts full path name of tar ball from the name of the .md5 file
+            # by trimming off the suffix.
             tb = Path(str(tbmd5)[0 : -len(".md5")])
-            tbdir = tb.parent
-            controller = tbdir.name
+            controller = tb.parent.name
 
-            try:
-                ProcessTb._results_push(controller, tb, self.token)
-            except Exception as e:
-                self.logger.error(
-                    "{}: Unexpected Error while running Agent's 'pbench-result-push' command: {}",
-                    self.config.TS,
-                    e,
-                )
-                nerr += 1
-                continue
+            if self.token:
+                # Step 1
+                # Only attempt to push a tar ball to a new Pbench Server via its
+                # PUT API when we have a token to use.  Any exceptions
+                # encountered are ignored s
+                try:
+                    ProcessTb._results_push(controller, tb, self.token)
+                except Exception as e:
+                    log_meth = (
+                        self.logger.warning
+                        if self.dispatch_states
+                        else self.logger.error
+                    )
+                    log_meth(
+                        "{}: Unexpected error while running Agent's 'pbench-result-push' command: {}",
+                        self.config.TS,
+                        e,
+                    )
+                    if not self.dispatch_states:
+                        # We have been asked to do one job, failures are errors
+                        # and we don't remove the failing tar ball so that the
+                        # PUT API can be retried.
+                        nerr += 1
+                        continue
 
-            try:
-                os.remove(tbmd5)
-            except Exception as exc:
-                self.logger.error(
-                    "{}: Cleanup of successful copy operation for MD5 file '{}' failed: '{}'",
-                    self.config.TS,
-                    tbmd5,
-                    exc,
-                )
-                nerr += 1
+            if not self.dispatch_states:
+                # We only have PUT API tasks cleanup the tar ball we just PUT.
+                nerr += self.remove(tbmd5, "cleanup of successful PUT API operation")
+                nerr += self.remove(tb, "cleanup of successful PUT API operation")
+            else:
+                assert False, f"unexpected {self.dispatch_states!r}"
```

Comment:
Wait, if we're going to `assert` out here, why didn't we just fail up front if no dispatch states were defined?